File: plateform/frontend/src/api/analytics.ts
import { apiClient } from "./client";
import type {
  AnalyticsAssetsResponse,
  AnalyticsBundle,
  AnalyticsDetectionsResponse,
  AnalyticsFilters,
  AnalyticsIncidentsResponse,
  AnalyticsOperationsResponse,
  AnalyticsOverviewResponse,
  AnalyticsTelemetryResponse,
  AnalyticsZonesResponse,
} from "../types/analytics";

const ANALYTICS_TIMEOUT = 20_000;

function toQuery(filters: AnalyticsFilters): Record<string, string> {
  const query: Record<string, string> = { range: filters.range };
  if (filters.zone.trim()) query.zone = filters.zone.trim();
  if (filters.sensor.trim()) query.sensor = filters.sensor.trim();
  return query;
}

export async function fetchAnalyticsOverview(
  filters: AnalyticsFilters,
  options?: { signal?: AbortSignal },
): Promise<AnalyticsOverviewResponse> {
  const { data } = await apiClient.get<AnalyticsOverviewResponse>("/api/analytics/overview", {
    params: toQuery(filters),
    signal: options?.signal,
    timeout: ANALYTICS_TIMEOUT,
  });
  return data;
}

export async function fetchAnalyticsBundle(
  filters: AnalyticsFilters,
  options?: { signal?: AbortSignal },
): Promise<AnalyticsBundle> {
  const params = toQuery(filters);
  const config = { params, signal: options?.signal, timeout: ANALYTICS_TIMEOUT };
  const [overview, telemetry, zones, detections, incidents, operations, assets] = await Promise.all([
    apiClient.get<AnalyticsOverviewResponse>("/api/analytics/overview", config),
    apiClient.get<AnalyticsTelemetryResponse>("/api/analytics/telemetry", config),
    apiClient.get<AnalyticsZonesResponse>("/api/analytics/zones", config),
    apiClient.get<AnalyticsDetectionsResponse>("/api/analytics/detections", config),
    apiClient.get<AnalyticsIncidentsResponse>("/api/analytics/incidents", config),
    apiClient.get<AnalyticsOperationsResponse>("/api/analytics/operations", config),
    apiClient.get<AnalyticsAssetsResponse>("/api/analytics/assets", config),
  ]);
  return {
    overview: overview.data,
    telemetry: telemetry.data,
    zones: zones.data,
    detections: detections.data,
    incidents: incidents.data,
    operations: operations.data,
    assets: assets.data,
  };
}

File: plateform/backend/app/api/routes/analytics.py
from fastapi import APIRouter, Depends, Query

from app.api.deps import get_analytics_service
from app.schemas.analytics import (
    AnalyticsAssetsResponse,
    AnalyticsDetectionsResponse,
    AnalyticsIncidentsResponse,
    AnalyticsOperationsResponse,
    AnalyticsOverviewResponse,
    AnalyticsTelemetryResponse,
    AnalyticsZonesResponse,
)
from app.services.analytics import AnalyticsService

router = APIRouter(prefix="/analytics", tags=["analytics"])


@router.get("/overview", response_model=AnalyticsOverviewResponse)
def get_analytics_overview(
    range: str = Query(default="24h"),
    zone: str | None = Query(default=None),
    sensor: str | None = Query(default=None),
    service: AnalyticsService = Depends(get_analytics_service),
) -> AnalyticsOverviewResponse:
    return service.overview(range, zone=zone, sensor=sensor)


@router.get("/telemetry", response_model=AnalyticsTelemetryResponse)
def get_analytics_telemetry(
    range: str = Query(default="24h"),
    zone: str | None = Query(default=None),
    sensor: str | None = Query(default=None),
    interval: str | None = Query(default=None),
    service: AnalyticsService = Depends(get_analytics_service),
) -> AnalyticsTelemetryResponse:
    return service.telemetry(range, zone=zone, sensor=sensor, interval=interval)


@router.get("/zones", response_model=AnalyticsZonesResponse)
def get_analytics_zones(
    range: str = Query(default="24h"),
    zone: str | None = Query(default=None),
    service: AnalyticsService = Depends(get_analytics_service),
) -> AnalyticsZonesResponse:
    return service.zones(range, zone=zone)


@router.get("/detections", response_model=AnalyticsDetectionsResponse)
def get_analytics_detections(
    range: str = Query(default="24h"),
    zone: str | None = Query(default=None),
    sensor: str | None = Query(default=None),
    interval: str | None = Query(default=None),
    service: AnalyticsService = Depends(get_analytics_service),
) -> AnalyticsDetectionsResponse:
    return service.detections(range, zone=zone, sensor=sensor, interval=interval)


@router.get("/incidents", response_model=AnalyticsIncidentsResponse)
def get_analytics_incidents(
    range: str = Query(default="24h"),
    zone: str | None = Query(default=None),
    sensor: str | None = Query(default=None),
    interval: str | None = Query(default=None),
    service: AnalyticsService = Depends(get_analytics_service),
) -> AnalyticsIncidentsResponse:
    return service.incidents(range, zone=zone, sensor=sensor, interval=interval)


@router.get("/operations", response_model=AnalyticsOperationsResponse)
def get_analytics_operations(
    range: str = Query(default="24h"),
    zone: str | None = Query(default=None),
    sensor: str | None = Query(default=None),
    interval: str | None = Query(default=None),
    service: AnalyticsService = Depends(get_analytics_service),
) -> AnalyticsOperationsResponse:
    return service.operations(range, zone=zone, sensor=sensor, interval=interval)


@router.get("/assets", response_model=AnalyticsAssetsResponse)
def get_analytics_assets(
    range: str = Query(default="24h"),
    zone: str | None = Query(default=None),
    sensor: str | None = Query(default=None),
    service: AnalyticsService = Depends(get_analytics_service),
) -> AnalyticsAssetsResponse:
    return service.assets(range, zone=zone, sensor=sensor)
